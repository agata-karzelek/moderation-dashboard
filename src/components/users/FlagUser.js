import React, { useState, useEffect } from "react";
import {
  DialogTitle,
  Dialog,
  Typography,
  IconButton,
  TextField,
  DialogActions,
  Button,
  CircularProgress,
  DialogContent,
} from "@material-ui/core";
import CloseIcon from "@material-ui/icons/Close";
import { useStyles } from "../../style/createModerationModal";
import { setUserMetadata } from "../../services/pubnub";
import SnackBar from "../core/SnackBar";
import { capitalizeFirstLetter, showError, selectedAccountsFromLS } from "../../utils/helpers";
import * as yup from "yup";
import { useFormik } from "formik";

const validationSchema = yup.object({
  reason: yup.string().required("Reason is required"),
});

export default function FlagUser(props) {
  const { pubnub, open, setOpen, uuid, isUpdated, action } = props;
  const [userID, setuserID] = useState("");
  const [flagAlert, setFlagAlert] = useState({
    success: { status: false, msg: "" },
    error: { status: false, msg: "" },
    loading: false,
  });
  const classes = useStyles();
  const selectedAcount = selectedAccountsFromLS();
  const updatedAt = new Date();
  const customMetaData = uuid.custom || {};

  useEffect(() => {
    setuserID(uuid.id);
  }, [uuid]);

  const handleClose = () => {
    formik.resetForm();
    setOpen(false);
  };

  const formik = useFormik({
    initialValues: {
      reason: "",
    },
    enableReinitialize: true,
    validationSchema: validationSchema,
    onSubmit: (values) => {
      flagUser(values);
    },
  });

  const flagUser = (values) => {
    setFlagAlert({
      ...flagAlert,
      success: { status: false, msg: "" },
      error: { status: false, msg: "" },
      loading: true,
    });
    customMetaData.reason = values.reason;
    customMetaData.flaggedBy = selectedAcount && selectedAcount.email;
    customMetaData.flaggedAt = updatedAt;
    customMetaData.flag = action === "flag" ? true : false;
    (async () => {
      try {
        const userResponse = await setUserMetadata(pubnub, userID, customMetaData);
        handleClose();
        setFlagAlert({
          ...flagAlert,
          success: {
            status: true,
            msg: `User ${capitalizeFirstLetter(action)}ged successfully.`,
          },
          error: { status: false, msg: "" },
          loading: false,
        });
        isUpdated(userResponse);
      } catch (e) {
        handleClose();
        setFlagAlert({
          ...flagAlert,
          success: { status: false, msg: "" },
          error: { status: true, msg: showError(e.status.errorData) },
          loading: false,
        });
      }
    })();
  };

  return (
    <>
      <Dialog
        fullWidth={true}
        onClose={handleClose}
        aria-labelledby="customized-dialog-title"
        open={open}
        className={classes.modalLayout}
      >
        <DialogTitle id="customized-dialog-title" onClose={handleClose} disableTypography>
          <Typography variant="h6" className={classes.headingFont}>
            {capitalizeFirstLetter(action)} User
          </Typography>
          <IconButton aria-label="close" onClick={handleClose} className={classes.closeButton}>
            <CloseIcon />
          </IconButton>
        </DialogTitle>
        <form onSubmit={formik.handleSubmit}>
          <DialogContent>
            <Typography testid="reason" className={classes.subHeadingFont}>
              Reason
            </Typography>
            <TextField
              id="reason"
              name="reason"
              placeholder="Add Reason"
              variant="outlined"
              className={classes.inputBox}
              autoComplete="off"
              onChange={formik.handleChange}
              value={formik.values.reason}
              error={formik.touched.reason && Boolean(formik.errors.reason)}
              helperText={formik.touched.reason && formik.errors.reason}
            />

            <br />
          </DialogContent>
          <DialogActions>
            <Button id="cancel" autoFocus className={classes.cancelButton} onClick={handleClose}>
              CANCEL
            </Button>

            <Button
              id="submit"
              autoFocus
              className={classes.addButton}
              type="submit"
              color="primary"
              disabled={flagAlert.loading}
              startIcon={
                flagAlert.loading ? <CircularProgress className={classes.loader} size={30} /> : null
              }
            >
              <Typography className={classes.buttonText}>Submit</Typography>
            </Button>
          </DialogActions>
        </form>
      </Dialog>
      {flagAlert.error.status && <SnackBar msg={flagAlert.error.msg} status={"info"} />}
      {flagAlert.success.status && <SnackBar msg={flagAlert.success.msg} status={"success"} />}
    </>
  );
}
