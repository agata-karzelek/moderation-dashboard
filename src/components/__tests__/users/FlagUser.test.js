import React from "react";
import { shallow } from "enzyme";
import FlagUser from "../../users/FlagUser";

const mockProp = {
  open: false,
  setOpen: jest.fn(),
  uuid: "",
  pubnub: {},
  action: "",
  //   children: ['You want to ', '', ' this user?'],
  isUpdated: jest.fn(),
};

describe("Test Cases for Flag User", () => {
  let component;
  beforeAll(() => {
    component = shallow(
      <FlagUser
        open={mockProp.open}
        setOpen={mockProp.setOpen}
        uuid={mockProp.uuid}
        isUpdated={mockProp.isUpdated}
        pubnub={mockProp.pubnub}
        action={mockProp.action}
      />
    );
  });

  test("Snapshhot", () => {
    expect(component).toMatchSnapshot();
  });

  test("initial Values", () => {
    expect(component.find('[testid="reason"]').text()).toBe("Reason");
    expect(component.find("#submit").text()).toBe("Submit");
    expect(component.find("#cancel").text()).toBe("CANCEL");
  });
});
