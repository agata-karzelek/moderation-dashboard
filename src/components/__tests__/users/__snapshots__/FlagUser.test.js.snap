// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Cases for Flag User Snapshhot 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Dialog))
    aria-labelledby="customized-dialog-title"
    className="makeStyles-modalLayout-8"
    fullWidth={true}
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(DialogTitle))
      disableTypography={true}
      id="customized-dialog-title"
      onClose={[Function]}
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-headingFont-13"
        variant="h6"
      >
         User
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(IconButton))
        aria-label="close"
        className="makeStyles-closeButton-3"
        onClick={[Function]}
      >
        <Memo(ForwardRef(CloseIcon)) />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(DialogTitle))>
    <form
      onSubmit={[Function]}
    >
      <WithStyles(ForwardRef(DialogContent))>
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-subHeadingFont-12"
          testid="reason"
        >
          Reason
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(TextField))
          autoComplete="off"
          className="makeStyles-inputBox-6"
          id="reason"
          name="reason"
          onChange={[Function]}
          placeholder="Add Reason"
          value=""
          variant="outlined"
        />
        <br />
      </WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogActions))>
        <WithStyles(ForwardRef(Button))
          autoFocus={true}
          className="makeStyles-cancelButton-9"
          id="cancel"
          onClick={[Function]}
        >
          CANCEL
        </WithStyles(ForwardRef(Button))>
        <WithStyles(ForwardRef(Button))
          autoFocus={true}
          className="makeStyles-addButton-14"
          color="primary"
          disabled={false}
          id="submit"
          startIcon={null}
          type="submit"
        >
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-buttonText-5"
          >
            Submit
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(DialogActions))>
    </form>
  </WithStyles(ForwardRef(Dialog))>
</Fragment>
`;
